staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -577,20 +601,41 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 			returnedRV = continueRV
 		}
 	case s.pagingEnabled && pred.Limit > 0:
-		if len(resourceVersion) > 0 {
-			fromRV, err := s.versioner.ParseResourceVersion(resourceVersion)
-			if err != nil {
-				return apierrors.NewBadRequest(fmt.Sprintf("invalid resource version: %v", err))
-			}
-			if fromRV > 0 {
-				options = append(options, clientv3.WithRev(int64(fromRV)))
+		if fromRV != nil {
+			switch match {
+			case metav1.ResourceVersionMatchNotOlderThan: // do nothing
+			case metav1.ResourceVersionMatchExact:
+				returnedRV = int64(*fromRV)
+				if returnedRV == 0 {
+					// Special case: exact + rv=0. etcd keyspace is empty. xref https://github.com/etcd-io/etcd/issues/11496
+					return s.versioner.UpdateList(listObj, 0, "", nil)
+				}
+				options = append(options, clientv3.WithRev(returnedRV))
+			case "": // legacy case
+				if *fromRV > 0 {
+					returnedRV = int64(*fromRV)
+					options = append(options, clientv3.WithRev(returnedRV))
+				}
 			}
-			returnedRV = int64(fromRV)
 		}
 
 		rangeEnd := clientv3.GetPrefixRangeEnd(keyPrefix)
 		options = append(options, clientv3.WithRange(rangeEnd))
 	default:
+		if fromRV != nil {
+			switch match {
+			case metav1.ResourceVersionMatchNotOlderThan: // do nothing
```

Comment:
rather than a `// do nothing` comment for this case and the legacy case, it might be clearer to:
1. set a local `ensureMinimumVersion = true` var
2. guard the call to `s.ensureMinimumResourceVersion` on that var rather than redoing the match check and hoping it that logic stays in sync with these switch cases

Reply:
Thanks. This works out to be quite a bit cleaner.